sycl/include/sycl/sycl.hpp

```diff
@@ -69,9 +69,13 @@
 #include <sycl/ext/oneapi/annotated_arg/properties.hpp>
 #include <sycl/ext/oneapi/device_global/device_global.hpp>
 #include <sycl/ext/oneapi/device_global/properties.hpp>
+#include <sycl/ext/oneapi/experimental/ballot_group.hpp>
 #include <sycl/ext/oneapi/experimental/bfloat16_math.hpp>
 #include <sycl/ext/oneapi/experimental/builtins.hpp>
+#include <sycl/ext/oneapi/experimental/cluster_group.hpp>
 #include <sycl/ext/oneapi/experimental/cuda/barrier.hpp>
+#include <sycl/ext/oneapi/experimental/opportunistic_group.hpp>
+#include <sycl/ext/oneapi/experimental/tangle_group.hpp>
```

Comment:
The compile time is not great already. Can we don't add these includes to `sycl.hpp` and ask users to include header files with experimental API explicitly?

Reply:
I'm not opposed to that, but would it be acceptable to do that in a separate PR?  There are some other experimental headers that I think we'd want to move (e.g. `cuda_barrier`) and some other headers that aren't obviously marked experimental but which implement experimental features (e.g. kernel properties).

I think it could be confusing if these new groups are the only experimental feature requiring users to include extra headers.

If we were to adopt this as a convention for experimental features, I think we'd need to update the documentation for our extensions too. So I'd like to hear @gmlueck's opinion about that.

Reply:
I suggest we remove all includes of headers with API extensions (not just experimental ones) and leave only headers with standard features API in `sycl.hpp`. I remember discussing this already, but I don't remember if we decided to clarify this in any documentation. On the other hand, as far as I know there are no guidelines to add includes for extension headers to standard `sycl.hpp` header neither.
I'm not sure if we do it easily for existing includes as it might be considered as API breaking change, but at least we can avoid doing it for new headers. Once breaking changes are allowed, we can clean-up `sycl.hpp` file.

> would it be acceptable to do that in a separate PR?

I think so, but, please, make sure doing this before any products based on intel/llvm are released. Otherwise, it's going to be an API breaking change and we usually keep API stable for a long period of time.

Reply:
It is definitely an API breaking change if we start requiring the user to `#include` a new header in order to use an extension.

I'd like to have a consistent strategy before we start changing things.  I agree with @Pennycook that it will be confusing if only a few extensions require the user to `#include`.  Possible options are:

* Require users to `#include` for experimental extensions but not supported or deprecated ones.
* Require users to `#include` for all extensions.  In this case, we need some plan to manage the transition.  For example, can we keep existing extensions as part of `sycl/sycl.hpp` with a deprecation warning?

Note that some extensions will not be able to be extracted into a separate header.  For example, extensions that add member functions to existing classes don't have a separate header.

What about extensions that add new types *and* also add new member functions to existing classes?  It might be weird to have the extension half supported when the user didn't `#include` the extension's header.

Maybe it would be better to have all the extensions implemented in `sycl/sycl.hpp`, but enable them via some command line option.  The option could predefine some macro which enables the extension code within the headers.

Reply:
> Maybe it would be better to have all the extensions implemented in sycl/sycl.hpp, but enable them via some command line option. The option could predefine some macro which enables the extension code within the headers.

Since every extension already defines a macro, could we just use those?

I could then update sycl.hpp to:
```c++
#ifdef SYCL_EXT_ONEAPI_NON_UNIFORM_GROUPS
#include <sycl/ext/oneapi/experimental/opportunistic_group.hpp>
#include <sycl/ext/oneapi/experimental/tangle_group.hpp>
#endif
```

(Or, alternatively, put that `#ifdef` inside each of the group headers.)

Then we could tell developers that to use experimental extensions they have to enable the macro themselves, whereas the macros for other (non-experimental) extensions are predefined?

> I'm not sure if we do it easily for existing includes as it might be considered as API breaking change, but at least we can avoid doing it for new headers. Once breaking changes are allowed, we can clean-up sycl.hpp file.

If we only make changes for experimental extensions, do breaking changes matter?

Reply:
Following [Zero-overhead_principle](https://en.cppreference.com/w/cpp/language/Zero-overhead_principle) "You don't pay for what you don't use" C++ standard library API is distributed across multiple headers: https://en.cppreference.com/w/cpp/header. Users include `<optional>` header to import `std::optional` class declaration. 

GNU libstdc++ implementation provides a non-standard `<bits/stdc++.h>` header file, which includes all the headers of C++ standard library. Using this header simplifies the use of C++ standard library functionality but incurs the overhead of parsing all the headers.

The SYCL specification defines a single header for the entire standard library/API. The spec doesn't define whether extensions should be defined in `sycl/sycl.hpp`, it just says that extensions can add new headers, and if they do, they must follow the naming scheme. The more features we add, the higher the overhead on parsing `sycl/sycl.hpp`.

Technically, the DPC++ library already provides separate headers for each class or feature, but we use them only for internal testing and don't promote using them to our users.

I believe that SYCL specification should provide users ability to include reduced set of declarations similar to standard C++ library. Having `sycl/sycl.hpp` including all standard declarations and extensions is fine as long as we have an alternative to avoid "usability overhead".

> Since every extension already defines a macro, could we just use those?

I don't think so. If feature doesn't require compiler support, the macro is usually defined in the header itself ([example](https://github.com/intel/llvm/blob/27654ac69fe26828ef7e5fb4a2a1bdac7dd1f86b/sycl/include/sycl/ext/oneapi/experimental/uniform.hpp#L14-L16)). The macro can be tested after include directive.

> Then we could tell developers that to use experimental extensions they have to enable the macro themselves, whereas the macros for other (non-experimental) extensions are predefined?

The macro is supposed to be set by implementation as it indicates whether extension is enabled by implementation. SYCL standard macro defined by user might be confusing. But we could have some controls to disable support for extensions/feature (e.g. compiler flag and/or `SYCL_EXT_DISABLE_***` macro for use before SYCL headers included).

We should move this discussion to `SYCL-Docs` project as we are not going to solve this problem in this PR, but I wanted to raise this issue to draw the attention to the compile time overhead problem.

Reply:
I'm 100% behind splitting SYCL out into multiple headers (or C++20 modules) to improve compile-times.  Any resistance I'm showing here is definitely self-serving: I'd like to get these initial class definitions merged so that we can move our attention to things like CUDA support and more complicated functionality (e.g. the algorithms).

You're right that the macro fix I proposed is probably just as confusing as requiring extra includes.  I agree that this is likely a bigger problem that we should discuss elsewhere... I suspect that it will take some time to reach broad consensus about what to do here, so what should I do to move this PR forward?

Reply:
I wish we don't put extension headers to `sycl.hpp` and teach users to include the right headers.
There are > 10 headers for `oneapi` experimental features, but only 3 of them snuck into `sycl.hpp`.

Having said that, this is not the most critical issue for this project, and we will need to address this problem for other features anyway. I won't block your patch, if you want to commit it with `sycl.hpp` changes.

Reply:
> I wish we don't put extension headers to `sycl.hpp` and teach users to include the right headers.
> There are > 10 headers for `oneapi` experimental features, but only 3 of them snuck into `sycl.hpp`.

This would be a good topic to discuss in our Friday "SYCL Language Discussion" meeting.  Would you like to summarize the current status of experimental extensions and their headers, and then present at the meeting?

Reply:
> > I wish we don't put extension headers to `sycl.hpp` and teach users to include the right headers.
> > There are > 10 headers for `oneapi` experimental features, but only 3 of them snuck into `sycl.hpp`.
> 
> This would be a good topic to discuss in our Friday "SYCL Language Discussion" meeting. Would you like to summarize the current status of experimental extensions and their headers, and then present at the meeting?

Unfortunately, I don't track DPC++ development closely enough to feel like I can present a summary of the current status. At the same time, I don't think the current status of experimental extensions and their headers is really important to the main topic of this discussion - compile time overhead. Experimental extensions are brought here as an example of uncontrolled compile time overhead accumulation because this PR implements an experimental extension and there is an easy way to avoid it (my POV).

There are multiple ways to deal with the compile time overhead problem and the SYCL specification can help with making the solution portable across multiple implementations. I will glad to join the discussion at the meeting if you send me invitation.